orchestrator/utils/errors.py

```diff
@@ -108,7 +108,7 @@ def is_api_exception(ex: Exception) -> bool:
     return ex.__class__.__name__ == "ApiException"
 
 
-def error_state_to_dict(err: Exception) -> ErrorDict:
+def error_state_to_dict(err: Union[Exception, ErrorDict]) -> ErrorDict:
```

Comment:
Haven't looked at the details, but cant this Union be avoided? Maybe having two separate functions? Also the doc string is not correct/complete anymore.

Reply:
This function is used as `step_result_process.on_failed(error_state_to_dict)` and it should just work on any `Process`. This line basically transforms a `Failure[Exception]` to `Failure[ErrorDict]`.

Then we have 2 versions of `Failure` in the code, one wrapping a dictionary and one wrapping an exception.
If you called this function on a `Failure` class wrapping a dictionary, it crashes. This has never happened because to there happened to be no code path where `step_result_process.on_failed(error_state_to_dict)` would be called on a `Failure[ErrorDict]`.

This update makes that operation idempotent. But I agree the signature is ugly. Preferably we should never carry exceptions around in `Failure` classes and serialize them immediately, but that is a bigger change.

Reply:
I could make it a singledispatch function to split up the separate behaviours based on the error type